### PR TITLE
Refactoring the binlog events library.

### PR DESCRIPTION
**Pull request by alainjobart**
Moving a few things around:
- mysqlctl/replication Status is moved up to mysqlctl.
- the rest of mysqlctl/replication is moved to a new library
  go/mysqlconn/replication.
- the binlog events for mysql56 and mariadb are also moved into that new
  library.
- Adding doc for new package.
- Adding support for RBR events in the low level library. Not fully plumbed in yet.

### CODE REVIEW

**Comment by alainjobart** [comment]
@sougou @michael-berlin this is progress for RBR. I want to be able to test replication from go/mysqlconn, so I need all these files.

**Comment by alainjobart** [comment]
@sougou I'll keep adding to this if you don't review it! ;)

(The RBR code is ugly and experimental, will need refactoring)

File: go/mysqlconn/conn.go [hunk 1]
```diff
@@ -256,7 +259,9 @@ func (c *Conn) writeOKPacket(affectedRows, lastInsertID uint64, flags uint16, wa
 	if err := c.writePacket(data); err != nil {
 		return err
 	}
-	c.flush()
+	if err := c.flush(); err != nil {
```
Comment (sougou) [inline go/mysqlconn/conn.go]: It's hard to reason about the correctness (and efficiency) of flush. If a write is always followed by a read, we could have the read perform the flush.
The other option would be to have the exported functions perform the flush.
Or both.

File: go/mysqlconn/replication/binlog_event.go [hunk 2]
```diff
@@ -0,0 +1,317 @@
+// Copyright 2014, Google Inc. All rights reserved.
```
Comment (sougou) [inline go/mysqlconn/replication/binlog_event.go]: 2017?

File: go/mysqlconn/replication/binlog_event.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+// Rows contains data from a {WRITE,UPDATE,DELETE}_ROWS_EVENT.
+type Rows struct {
+	// Flags has the flags from the event.
+	Flags uint16
+
+	// IdentifyColumns describes which columns are included to
+	// identify the row. It is a bitmap indexed by the TableMap
+	// list of columns.
+	// Set for UPDATE and DELETE.
+	IdentifyColumns Bitmap
```
Comment (sougou) [inline go/mysqlconn/replication/binlog_event.go]: Is there a benefit to these being a bitmap, vs. a []bool?

File: go/mysqlconn/replication/constants.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
+package replication
+
+// This file contains the constant definitions for this package.
+
+// This is the data type for a field.
+// Values taken from include/mysql/mysql_com.h
+// FIXME(alainjobart) this is the same info as constants describes in
```
Comment (sougou) [inline go/mysqlconn/replication/constants.go]: sqltypes.Value also has some duplication that will require consolidation.
I also think the ones in mysql.go are unused now.

File: go/mysqlconn/conn.go (hunk 1 above)
Comment (alainjobart) [inline go/mysqlconn/conn.go]: Actually, for correctness, the unit tests catch it right away: no flush, the other side doesn't get the packet (happened a few times during development). So I don't think that's much of a concern.

File: go/mysqlconn/replication/binlog_event.go (hunk 3 above)
Comment (alainjobart) [inline go/mysqlconn/replication/binlog_event.go]: Yes, no memory copy, no allocation. We just point to the data of the original packet.

File: go/mysqlconn/replication/constants.go (hunk 4 above)
Comment (alainjobart) [inline go/mysqlconn/replication/constants.go]: Agreed.

File: go/mysqlconn/replication/binlog_event.go (hunk 2 above)
Comment (alainjobart) [inline go/mysqlconn/replication/binlog_event.go]: Will fix them all at once in a commit.

**Pull request merged**